k2/csrc/ragged_ops.h

```diff
@@ -804,18 +822,88 @@ RaggedShape RemoveEmptyListsAxis0(RaggedShape &src_shape,
+    @param [out] elems_new2old   If supplied, this function will
+                    output to this location a new2old array that
+                    dictates how the elements of a ragged tensor
+                    with shape `src` would be renumbered.
+    @return  Returns the subsampled shape. All dimensions and tot-sizes
+       preceding the final axis will remain the same, which might give
+       rise to empty lists on those axes; these can be removed if
+       necessary with RemoveEmptyLists().
  */
 template <typename T>
-Ragged<T> SubsampleRagged(Ragged<T> &src, Renumbering &renumbering) {
-  return Ragged<T>(SubsampleRaggedShape(src.shape, renumbering),
-                   src.values[renumbering.New2Old()]);
+Ragged<T> SubsetRagged(Ragged<T> &src, Renumbering &renumbering,
+                       int32_t axis = -1,
+                       Array1<int32_t> *elems_new2old = nullptr) {
+  Array1<int32_t> tmp;
+  if (elems_new2old == nullptr)
+    elems_new2old = &tmp;
+  RaggedShape shape = SubsetRaggedShape(src.shape, renumbering,
+                                           axis, elems_new2old);
+  return Ragged<T>(shape, src.values[*elems_new2old]);
 }
 
+/*
+  This function creates a Renumbering object that can be used to obtain subsets
+  of ragged arrays via SubsetRaggedShape().  It implements beam pruning as
+  used in pruned Viterbi search and similar algorithms, where there is both a
+  beam and a max-active (`max_elems`) constraint.  T will probably be float or
+  double, interpreted as a "positive-is-better" sense, i.e. as scores.
+
+   @param [in] src  The ragged object to be subsampled.
+   @param [in] axis  The axis to be subsampled, must satisfy
+              0 <= axis < src.NumAxes().  The axis before `axis`, if axis < 0,
+              will be interpreted as a "batch" axis.
+   @param [in] beam  The main pruning beam.  The sub-lists of elements on axis
+              `axis` will be removed if their maximum element (or the element
+              itself, if axis + 1 == src.NumAxes()) is less than
+              this_best_elem - beam, where this_best_elem
+              is the maximum element taken over axis `axis-1` (or over the
+              entire array, if axis == 0).   Think of axis `axis-1`, if
+              present, as the "batch" axis, and axis `axis` as the axis that we
+              actually remove elements or sub-lists on.  Empty sub-lists on axis
+              `axis` will always be pruned, as their score would be treated
+              as -infinity.
+   @param [in] max_elems  If max_elems > 0, it is the maximum number of sub-lists
+              or elements that are allowed within any sub-list on axis `axis-1`
+              (or the maximum number of top-level sub-lists after subsampling,
+              if axis == 0).  We keep the best ones, but behavior in case of ties is
+              undefined (TODO: check whether SortSublists() is a stable sort, and
```

Comment:
please resolve this TODO, we need to clarify if the behavior, if there are ties, is well-defined or could be random.

Reply:
I just checked the code, when running on CPU we use `std::sort` which is stable with elements number less than 16 (algorithm is insert sort) and is unstable with elements number equals to or greater than 16 (algorithm is quicksort).
when running on GPU we use the `segmented_sort` in modern_gpu lib which is modified from merge sort, and it is stable. So do we need to clarify the different behavior of different devices, especially the 16 elements thing?

Reply:
Maybe, we can change the cpu sorting algorithm from `std::sort` to `std::stable_sort`.

Reply:
Sure, I guess that is OK.